### PR TITLE
M3: Refactor AbTestHelper in all bundles

### PR DESCRIPTION
**Pull request by ondrejsibl**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Automated tests included? |Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/8067
| BC breaks? | AbTestHelper class removed, use DetermineWinnerSubscriber instead
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Refactor AbTestHelper in all bundles to get rid of static methods and use events instead. The goal is to get rid of MauticFactory.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create A/B tests for Email and Page. Verify, that all winner criteria work like before and winner is being chosen (seems we can't use submissions for emails, because we can't use forms in emails. Also, in bounce rate criteria for landing pages, we are choosing the highest rate as a winner instead of the lowest one, maybe even the rate calculation is not right. But that is not a part of this task).

### CODE REVIEW

File: app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php [hunk 1]
```diff
@@ -1,37 +1,66 @@
 <?php
 
 /*
- * @copyright   2014 Mautic Contributors. All rights reserved
+ * @copyright   2019 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php]: Do not change this for existing files. It should be the current year only for newly created files.
```suggestion
 * @copyright   2014 Mautic Contributors. All rights reserved
```

File: app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php [hunk 2]
```diff
@@ -1,37 +1,66 @@
...
-     * @param               $children
-     *
-     * @return array
+     * @param DetermineWinnerEvent $event
      */
-    public static function determineDownloadWinner($factory, $parent, $children)
+    public function onDetermineDownloadRateWinner(DetermineWinnerEvent $event)
     {
-        $repo = $factory->getEntityManager()->getRepository('MauticAssetBundle:Download');
+        $repo       = $this->em->getRepository('MauticAssetBundle:Download');
```
Comment (escopecz) [inline app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php]:
```suggestion
        $repo       = $this->em->getRepository(Download::class);
```
+ add the use statement of the Download entity.

File: app/bundles/CoreBundle/Event/DetermineWinnerEvent.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * @var array
+     */
+    private $abTestResults;
+
+    /**
+     * @param array $parameters
+     */
+    public function __construct($parameters)
```
Comment (escopecz) [inline app/bundles/CoreBundle/Event/DetermineWinnerEvent.php]:
```suggestion
    public function __construct(array $parameters)
```

File: app/bundles/CoreBundle/Event/DetermineWinnerEvent.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public function getAbTestResults()
+    {
+        return $this->abTestResults;
+    }
+
+    /**
+     * @param array $abTestResults
+     */
+    public function setAbTestResults($abTestResults)
```
Comment (escopecz) [inline app/bundles/CoreBundle/Event/DetermineWinnerEvent.php]:
```suggestion
    public function setAbTestResults(array $abTestResults)
```

File: app/bundles/EmailBundle/EventListener/DetermineWinnerSubscriber.php [hunk 5]
```diff
@@ -1,33 +1,69 @@
 <?php
 
 /*
- * @copyright   2014 Mautic Contributors. All rights reserved
+ * @copyright   2019 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/DetermineWinnerSubscriber.php]:
```suggestion
 * @copyright   2014 Mautic Contributors. All rights reserved
```

File: app/bundles/FormBundle/EventListener/DetermineWinnerSubscriber.php [hunk 6]
```diff
@@ -1,37 +1,66 @@
 <?php
 
 /*
- * @copyright   2014 Mautic Contributors. All rights reserved
+ * @copyright   2019 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline app/bundles/FormBundle/EventListener/DetermineWinnerSubscriber.php]:
```suggestion
 * @copyright   2014 Mautic Contributors. All rights reserved
```

File: app/bundles/PageBundle/EventListener/DetermineWinnerSubscriber.php [hunk 7]
```diff
@@ -1,37 +1,66 @@
 <?php
 
 /*
- * @copyright   2014 Mautic Contributors. All rights reserved
+ * @copyright   2019 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline app/bundles/PageBundle/EventListener/DetermineWinnerSubscriber.php]:
```suggestion
 * @copyright   2014 Mautic Contributors. All rights reserved
```

File: app/bundles/FormBundle/EventListener/DetermineWinnerSubscriber.php (hunk 6 above)
Comment (ondrejsibl) [reply app/bundles/FormBundle/EventListener/DetermineWinnerSubscriber.php]: This is a new file, but it is based on an old one, so github sees it as a change.

File: app/bundles/PageBundle/EventListener/DetermineWinnerSubscriber.php (hunk 7 above)
Comment (ondrejsibl) [reply app/bundles/PageBundle/EventListener/DetermineWinnerSubscriber.php]: This is a new file, but it is based on an old one, so github sees it as a change.

File: app/bundles/EmailBundle/EventListener/DetermineWinnerSubscriber.php (hunk 5 above)
Comment (ondrejsibl) [reply app/bundles/EmailBundle/EventListener/DetermineWinnerSubscriber.php]: This is a new file, but it is based on an old one, so github sees it as a change.

File: app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php (hunk 1 above)
Comment (ondrejsibl) [reply app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php]: This is a new file, but it is based on an old one, so github sees it as a change.

File: app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php (hunk 2 above)
Comment (ondrejsibl) [reply app/bundles/AssetBundle/EventListener/DetermineWinnerSubscriber.php]: Done.

File: app/bundles/CoreBundle/Event/DetermineWinnerEvent.php (hunk 4 above)
Comment (ondrejsibl) [reply app/bundles/CoreBundle/Event/DetermineWinnerEvent.php]: Done.

File: app/bundles/CoreBundle/Event/DetermineWinnerEvent.php (hunk 3 above)
Comment (ondrejsibl) [reply app/bundles/CoreBundle/Event/DetermineWinnerEvent.php]: Done.

**Comment by escopecz** [comment]
Works as described 👍

**Pull request merged**